src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -5437,6 +5462,1076 @@ class CL1<T>
+using System.Runtime.CompilerServices;
+class C
+{
+    void Main(string? s)
+    {
+        ThrowIfNull(42, s);
+        s.ToString(); // ok
+    }
+    [System.Diagnostics.Conditional(""DEBUG"")]
+    static void ThrowIfNull(int x, [EnsuresNotNull] string? s) => throw null;
+}
+" + EnsuresNotNullAttributeDefinition, parseOptions: TestOptions.Regular8, options: TestOptions.ReleaseDll);
+
+            c.VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void EnsuresNotNull_SecondArgumentDereferences()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+using System.Runtime.CompilerServices;
+public class C
+{
+    void Main(string? s)
+    {
+        ThrowIfNull(s, s.ToString()); // warn
+        s.ToString(); // ok
+    }
+    public static void ThrowIfNull([EnsuresNotNull] string? s, string s2) => throw null;
+}
+" + EnsuresNotNullAttributeDefinition, parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics(
+                // (7,24): warning CS8602: Possible dereference of a null reference.
+                //         ThrowIfNull(s, s.ToString()); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(7, 24)
+                );
+
+            VerifyAnnotations(c, "C.ThrowIfNull", EnsuresNotNull, None);
+        }
+
+        [Fact]
+        public void EnsuresNotNull_SecondArgumentAssigns()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+using System.Runtime.CompilerServices;
+class C
+{
+    void Main(string? s)
+    {
+        ThrowIfNull(s, s = null);
+        s.ToString(); // warn
+    }
+    static void ThrowIfNull([EnsuresNotNull] string? s, string? s2) => throw null;
+}
+" + EnsuresNotNullAttributeDefinition, parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics(
+                // (8,9): warning CS8602: Possible dereference of a null reference.
+                //         s.ToString(); // warn
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "s").WithLocation(8, 9)
+                );
+        }
+
+        [Fact]
+        public void EnsuresNotNull_String_Contains()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+class C
+{
+    void Main(string? s)
+    {
+        ""hello"".Contains(s);
+        s.ToString(); // ok
+    }
+}
+", parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics();
+
+            VerifyAnnotations(c, "System.String.Contains", EnsuresNotNull);
+        }
+
+        [Fact]
+        public void EnsuresNotNull_Indexer()
+        {
+            CSharpCompilation c = CreateCompilation(@"
+using System.Runtime.CompilerServices;
+class C
+{
+    void Main(string? s)
+    {
+        _ = this[42, s];
+        s.ToString(); // ok
+    }
+    public int this[int x, [EnsuresNotNull] string? s] => throw null;
+}
+" + EnsuresNotNullAttributeDefinition, parseOptions: TestOptions.Regular8);
+
+            c.VerifyDiagnostics();
+        }
```

Comment:
Consider testing:
```
static void EnsuresNotNull([EnsuresNotNull] params object[] args) { }
static void F(object? x, object? y, object[]? a)
{
    EnsuresNotNull(a);
    a.ToString(); // no warning
    EnsuresNotNull(x, y);
    x.ToString(); // warning
    y.ToString(); // warning
}

Reply:
Consider testing:
```
interface I
{
    void F([EnsuresNotNull] object? o);
}
delegate void D([EnsuresNotNull] object? o);
```

Reply:
Good catch. There is a bug on params (the two warnings aren't produced).

---
In reply to: [187462263](https://github.com/dotnet/roslyn/pull/26656#discussion_r187462263) [](ancestors = 187462263)